### PR TITLE
Fix banner and voice instructions observers reactive behavior

### PR DESCRIPTION
**Pull request by Guardiola31337**
### Description
Fixes banner and voice instructions observers reactive behavior

`MapboxNavigationActivity` registers the observers when clicking on start navigation button 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/3726d6a08636b77648da7f2e7e545d6b27865139/test-app/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt#L471-L473 and at that point the `bannerInstructionEvent.bannerInstructions` is `null` because the non-null `bannerInstructions` from NN was sent previously so the `BannerInstructionsObserver` callback is not fired-off

https://github.com/mapbox/mapbox-navigation-android/blob/3726d6a08636b77648da7f2e7e545d6b27865139/test-app/src/main/java/com/mapbox/navigation/examples/core/MapboxNavigationActivity.kt#L262-L266

causing a missing banner instruction 👀 

https://user-images.githubusercontent.com/1668582/109816381-aaa85400-7bfe-11eb-9688-9bcf10b2e876.mp4

https://user-images.githubusercontent.com/1668582/109816451-bf84e780-7bfe-11eb-81c2-2137d12bde30.mp4

Now we keep checking if there are new non-null `bannerInstructions` 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/337ccf0b3a9d28444792a1b96d39a97765bcf142/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L538

and if so emitting the callback 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/337ccf0b3a9d28444792a1b96d39a97765bcf142/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L539-L543

but when registering, the event is sent if there's a non-null value available 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/337ccf0b3a9d28444792a1b96d39a97765bcf142/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L333-L335

https://github.com/mapbox/mapbox-navigation-android/blob/337ccf0b3a9d28444792a1b96d39a97765bcf142/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L556

Same thing applies to `voiceInstructions`.

This has also surfaced an issue (regression?) in NN as supposedly these values should be populated (non-null) from NN only when updated but if you stand still this is being sent multiple times. As seen in the above _after_ video it's pretty noticeable as announcements are repeated over and over until you start moving. I also confirmed that with logs for `bannerInstructions`. @mskurydin @SiarheiFedartsou could you take a look at it when you have a chance? 🙏 

### Changelog
```
<changelog>Fixed banner and voice instructions observers reactive behavior.</changelog>
```

cc @etl @mskurydin @SiarheiFedartsou

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 1]
```diff
@@ -553,34 +551,30 @@ internal class MapboxTripSession(
...
-                val bannerComponents = bannerView?.components()
-                ifNonNull(bannerComponents) { components ->
-                    components.forEachIndexed { index, component ->
-                        component.takeIf { it.type() == BannerComponents.GUIDANCE_VIEW }?.let { c ->
-                            components[index] =
-                                c.toBuilder()
-                                    .imageUrl(c.imageUrl()?.plus("&access_token=$accessToken"))
-                                    .build()
-                        }
+        ifNonNull(bannerInstructionEvent.bannerInstructions) { bannerInstructions ->
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:556]: Supposedly this should be populated (non-null) from NN only when updated but if you stand still (stale location) this is being sent multiple times. Can be easily tested in `MapboxNavigationActivity`. It seems a regression we introduced at some point. Could you take a look at it @mskurydin @SiarheiFedartsou when you have a chance? 🙏

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 2]
```diff
@@ -553,34 +551,30 @@ internal class MapboxTripSession(
...
         }
     }
 
     private fun checkVoiceInstructionEvent(
-        progress: RouteProgress,
         action: (VoiceInstructions) -> Unit
     ) {
-        if (voiceInstructionEvent.isOccurring(progress)) {
-            action(voiceInstructionEvent.voiceInstructions)
+        ifNonNull(voiceInstructionEvent.voiceInstructions) { voiceInstructions ->
```
Comment (Guardiola31337) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:576]: Supposedly this should be populated (non-null) from NN only when updated but if you stand still this is being sent multiple times. Can be easily tested in `MapboxNavigationActivity`. It's pretty noticeable here as announcements are repeated over and over until you start moving. It seems a regression we introduced at some point. Could you take a look at it @mskurydin @SiarheiFedartsou when you have a chance? 🙏

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 3]
```diff
@@ -539,12 +535,14 @@ internal class MapboxTripSession(
         tripService.updateNotification(progress)
         progress?.let {
             routeProgressObservers.forEach { it.onRouteProgressChanged(progress) }
-            checkBannerInstructionEvent(progress) { bannerInstruction ->
+            bannerInstructionEvent.isOccurring(progress)
+            checkBannerInstructionEvent { bannerInstruction ->
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:539]: Could you explain a little bit more what and why we're trying to fix?

As far as I understand, we don't anymore check if the banner instructions are in the progress, we will always append the latest.

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:539]: I've just updated the description. Let me know if it makes sense.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:539]: Thanks! But now we're ignoring the result of `bannerInstructionEvent.isOccurring(progress)`, so we will push banner/voice instructions with each progress update, won't we?

If there are no new updates, we will push the old ones again. Or am I missing something?

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:543]: Great catch! Addressed in the latest commit pushed.

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:543]: If I'm not mistaken, we still have the same issue, don't we?

Let's imagine this scenario:
1. Route progress with the banner is generated (no observer, developer is not notified).
2. Developer registers observer and is notified.
3. Route progress **without** banner is generated (observer is already registered and the developer is notified again with the old value, `bannerInstructionEvent.latestBannerInstructions` is not null).

We should tie the `latestBannerInstructions` only to the moment when the developer registers the listener.

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:543]: We're actually doing that 👀 

https://github.com/mapbox/mapbox-navigation-android/blob/3ddd3c64f55988d0b6f9a8b8a0b9ccb4c57cc960/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L333

https://github.com/mapbox/mapbox-navigation-android/blob/3ddd3c64f55988d0b6f9a8b8a0b9ccb4c57cc960/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L557-L563

Same thing for `voiceInstructions`. Am I missing something?

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:543]: Okay this is resolved.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 4]
```diff
@@ -539,48 +539,74 @@ internal class MapboxTripSession(
...
+        action: (BannerInstructions) -> Unit
+    ) {
+        ifNonNull(bannerInstructionEvent.latestBannerInstructions) {
+            action(it)
+        }
+    }
+
+    private fun invalidateLatestBannerInstructionEvent() {
+        if (bannerInstructionsObservers.isEmpty()) {
+            bannerInstructionEvent.invalidateLatestBannerInstructions()
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:567]: I think this should be called when the route changes, not when observers are unsubscribed, shouldn't it?

**Review by LukasPaczos — COMMENTED** [review]
I understand the need for caching, but it can introduce an unexpected behavior, especially for voice:
1. The user is driving and the developer does not have a voice observer registered.
2. NN produces a voice update, "turn right in 700 meters".
3. User keeps driving.
4. Developer registers the voice observer and plays the cached "turn right in 700 meters" even though the user is now only 200 meters away from the maneuver.

In this case, it might've been better to avoid playing anything instead of playing an outdated message. Should we instead cache only banner instructions and not voice?

/cc @jyrigo

**Comment by Guardiola31337** [comment]
This is a great point. Although this is less likely to happen because voice events are normally sent at a later time (as opposed to banner instructions that are sent right away) I'm going to remove it for now.

cc @jyrigo

**Comment by Guardiola31337** [comment]
@LukasPaczos updated the PR. This is ready for another round of eyes 👀

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt [hunk 5]
```diff
@@ -363,7 +360,7 @@ internal class MapboxTripSession(
     ) {
         voiceInstructionsObservers.add(voiceInstructionsObserver)
         routeProgress?.let {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:362]: This check isn't needed anymore, is it?

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt [hunk 6]
```diff
@@ -2,19 +2,16 @@ package com.mapbox.navigation.core.trip.session
 
 import com.mapbox.api.directions.v5.models.VoiceInstructions
 import com.mapbox.navigation.base.trip.model.RouteProgress
-import com.mapbox.navigation.utils.internal.ifNonNull
 
 internal class VoiceInstructionEvent {
```
Comment (LukasPaczos) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt:6]: Looks like this class is not needed anymore, is it?

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+import com.mapbox.api.directions.v5.models.BannerInstructions
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import io.mockk.every
+import io.mockk.mockk
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNull
+import org.junit.Test
+
+class BannerInstructionEventTest {
```
Comment (LukasPaczos) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt:11]: This class is missing a test for:
1. Non-null banner instruction provided.
2. Null banner instruction provided.
3. `assertEquals(anyBannerInstructions, bannerInstructionEvent.latestBannerInstructions)`

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt (hunk 5 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt:362]: It's unlikely to be `true` but technically feasible. I removed it in the first place and then thought about it and decided it to leave it as we had it before.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt (hunk 6 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt:6]: Yes it is. We need to keep updating the voice instructions 👀

https://github.com/mapbox/mapbox-navigation-android/blob/9abce409b2d7f9f564dd6e29f102ad9bc0bc5f2b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L545-L546

https://github.com/mapbox/mapbox-navigation-android/blob/9abce409b2d7f9f564dd6e29f102ad9bc0bc5f2b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L586-L592

Why did you think is not needed anymore? 🤔

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt (hunk 7 above)
Comment (Guardiola31337) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt:11]: As it's always updated, isn't that covered with

https://github.com/mapbox/mapbox-navigation-android/blob/9abce409b2d7f9f564dd6e29f102ad9bc0bc5f2b/libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt#L14

and 

https://github.com/mapbox/mapbox-navigation-android/blob/9abce409b2d7f9f564dd6e29f102ad9bc0bc5f2b/libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt#L26

🤔

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt (hunk 6 above)
Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt:6]: It only checks nullability, why couldn't we do
```
            ifNonNull(status.voiceInstruction) { voiceInstruction ->
                voiceInstructionsObservers.forEach {
                    it.onNewVoiceInstructions(voiceInstruction)
                }
            }
```

Why do we need the whole `VoiceInstructionEvent` class?

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt (hunk 7 above)
Comment (LukasPaczos) [reply libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt:11]: It's not as far as I can tell.

We are not verifying that `latestBannerInstructions` is not cleared by a `null` banner instructions.

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt (hunk 6 above)
Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt:6]: We're also doing https://github.com/mapbox/mapbox-navigation-android/blob/9abce409b2d7f9f564dd6e29f102ad9bc0bc5f2b/libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt#L545 plus wanted to have consistency with `BannerInstructionEvent` ¯\\\_(ツ)\_/¯

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt:6]: The logic is unnecessarily complicated in my opinion. if we don't need that separate class, let's remove it.

Comment (Guardiola31337) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt:6]: It's also helping testing the logic in isolation (separation of concerns).

Comment (LukasPaczos) [reply libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/VoiceInstructionEvent.kt:6]: We're effectively extracting a null-check to a separate class. This isn't worth it in my opinion and makes the logic harder to follow.

**Comment by Guardiola31337** [comment]
Addressed the feedback @LukasPaczos Ready for another round 🚀

File: libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    @Test
+    fun `isOccurring doesn't update current latestBannerInstructions if instructions null`() {
+        val bannerInstructionEvent = BannerInstructionEvent()
+        val mockedNonNullBannerInstructionsRouteProgress: RouteProgress = mockk()
+        val nonNullBannerInstructions: BannerInstructions = mockk()
+        val nullBannerInstructions: BannerInstructions? = null
+        every {
+            mockedNonNullBannerInstructionsRouteProgress.bannerInstructions
+        } returns nonNullBannerInstructions andThen nullBannerInstructions
```
Comment (LukasPaczos) [inline libnavigation-core/src/test/java/com/mapbox/navigation/core/trip/session/BannerInstructionEventTest.kt:69]: woah, first time seeing this syntax :O

**Review by LukasPaczos — APPROVED** [review]
Thanks for addressing the comments, :ship:

**Pull request merged**